src/lib/OAuth2/TestOAuth2.ts

```diff
@@ -0,0 +1,163 @@
+
+  public async logout(): Promise<void> {
+    this.loggedInUser = null;
+    this.dispatchEvent(OAuth2Events.UserSignedOut);
+
+    return Promise.resolve();
+  }
+
+  public addEventListener<
+    T extends OAuth2Events,
+    U extends IOAuth2EventCallbacks[T]
+  >(event: T, cb: U) {
+    this.callbacks[event].push(cb);
+  }
+
+  public removeEventListener<
+    T extends OAuth2Events,
+    U extends IOAuth2EventCallbacks[T]
+  >(event: T, fn: U) {
+    this.callbacks[event].filter((cb) => !Object.is(cb, fn));
+  }
+
+  public expireUser() {
+    if (!this.loggedInUser) return;
+
+    const expiredDate = new Date(0);
+    this.loggedInUser = TestOAuth2.createLoggedInUser(expiredDate, expiredDate);
+    this.dispatchEvent(OAuth2Events.TokenExpired);
+    this.dispatchEvent(OAuth2Events.UserLoaded, this.loggedInUser);
+  }
+
+  public dispatchEvent<T extends OAuth2Events, P>(event: T, payload?: P) {
+    let dispatchedEvent: CustomEvent | null = null;
+    if (payload) {
+      dispatchedEvent = new CustomEvent(event, {
+        detail: payload,
+      });
+    }
+
+    for (const cb of this.callbacks[event]) {
+      cb(dispatchedEvent!);
+    }
+  }
+
+  public loggedInUser: IOAuth2User | null = null;
+
+  public callbacks: Record<
+    keyof IOAuth2EventCallbacks,
+    IOAuth2EventCallbacks[keyof IOAuth2EventCallbacks][]
+  > = {
+    [OAuth2Events.UserLoaded]: [],
+    [OAuth2Events.TokenExpired]: [],
+    [OAuth2Events.TokenExpiring]: [],
+    [OAuth2Events.UserUnloaded]: [],
+    [OAuth2Events.UserSignedOut]: [],
+    [OAuth2Events.SilentRenewError]: [],
+  };
+
+  public static createLoggedInUser(
+    creationDate = new Date(),
+    expirationDate = add({ days: 1 })(creationDate)
+  ): IOAuth2User {
+    // eslint-disable-next-line no-magic-numbers
+    const nowDuration = Math.ceil(Number(creationDate) / 1000); // in seconds.
+
+    // eslint-disable-next-line no-magic-numbers
+    const expirationDuration = Math.ceil(Number(expirationDate) / 1000); // in seconds.
+
+    const groups = ['some-group'];
+    const email = 'developer@giantswarm.io';
+    const emailVerified = true;
+
+    const header = {
+      alg: 'RS256',
+      kid: 'some-random-key',
+    };
+
+    const tokenPayload = {
+      iss: 'https://not-so-quick.com',
+      sub: 'a-key',
+      aud: ['cool'],
+      exp: expirationDuration,
+      iat: nowDuration,
+      azp: 'some-other-key',
+      at_hash: 'lolz',
+      email,
+      email_verified: emailVerified,
+      groups,
+      name: 'Developer McDeveloperson',
+      preferred_username: 'mcDeveloperson',
+    };
+
+    const signature = 'haha-not-so-fast';
+
+    const encode = (part: string | Record<string, unknown>) => {
+      const asBase64 = window.btoa(JSON.stringify(part));
+
+      return asBase64.replace(/-/g, '+').replace(/_/g, '/');
+    };
+
+    const idToken = `${encode(header)}.${encode(tokenPayload)}.${encode(
+      signature
+    )}`;
+
+    return {
+      refreshToken: 'some-random-key',
```

Comment:
I guess we're storing the refresh token in the local storage then? That's a bit risky since I think it never expires.

we should look into if Dex supports the PKCE flow, that'll also make it so we don't need a clientSecret here

Reply:
Sweet. Does that mean there might be a issue with running happa in multiple tabs?